File: 01_basic/week4/819_most-common-word_watermelon.py
class Solution:
    def mostCommonWord(self, paragraph: str, banned: List[str]) -> str:
        # 단어는 case-insensitive하므로 계산 편의를 위해 미리 소문자로 변환
        paragraph = paragraph.lower()
        
        # 영문자와 공백을 제외한 나머지 기호를 공백으로 대치
        # 처음엔 빈 문자열로 대치했다가 "a,b,c"와 같은 경우에서 틀림
        paragraph = re.sub('[^a-zA-Z\s]', ' ', paragraph)
        
        words = paragraph.split()  # 단어 저장 리스트
        words_count = dict()  # 단어 등장 횟수 저장 딕셔너리

        most_common_count = 0  # 한 단어가 가장 많이 등장한 횟수
        most_common_word = ''  # 가장 많이 등장한 단어
        
        banned = set(banned)  # 빠른 탐색을 위해 리스트를 세트로 변환
        
        for word in words:
            if word in banned:  # 단어가 금지된 단어 세트 안에 있다면 무시
                continue
            
            if word in words_count:  # 단어가 딕셔너리에 기록된 적 있는 경우
                words_count[word] += 1  # 등장 횟수 갱신
                    
            else:  # 단어가 딕셔너리에 기록된 적 없는 경우 새로 기록
                words_count[word] = 1
                
            # 최대 등장 횟수를 넘는 경우 횟수와 단어 갱신
            if most_common_count < words_count[word]:
                most_common_count = words_count[word]
                most_common_word = word
                
        return most_common_word
                

File: 01_basic/week4/819_most-common-word_watermelon.js
/**
 * @param {string} paragraph
 * @param {string[]} banned
 * @return {string}
 */
var mostCommonWord = function(paragraph, banned) {
    paragraph = paragraph.toLowerCase();

    // g: 대상 문자열 내에서 패턴과 일치하는 모든 문자열을 전역 검색
    paragraph = paragraph.replace(/[^a-zA-Z\s]/g, ' ');
    
    const words = paragraph.split(/\s+/g);
    const wordsCount = {};
    let mostCommonCount = 0;
    let answer = '';
    
    const bannedSet = new Set(banned);
    
    // split 결과로 맨 뒤에 빈 문자열이 생길 수 있으므로 이를 bannedSet에 추가
    bannedSet.add('')  
    
    for (word of words) {
        if (bannedSet.has(word)) continue;
        
        // wordsCount 안에 word가 존재한다면 그 값에서 1만큼 더 큰 수를,
        // 존재하지 않는다면 1을 할당
        wordsCount[word] = wordsCount[word] ? wordsCount[word] + 1 : 1;
        
        if (mostCommonCount < wordsCount[word]) {
            mostCommonCount = wordsCount[word];
            answer = word;
        }
    }
    
    return answer;
};
